exploration/0005-open-close-expressions.md

```diff
@@ -0,0 +1,152 @@
+# Open/Close Expressions
+
+Status: **Proposed**
+
+<details>
+	<summary>Metadata</summary>
+	<dl>
+		<dt>Contributors</dt>
+		<dd>@eemeli</dd>
+		<dd>@aphillips</dd>
+		<dt>First proposed</dt>
+		<dd>2023-09-05</dd>
+		<dt>Pull Request</dt>
+		<dd><a href="https://github.com/unicode-org/message-format-wg/pull/470">#470</a></dd>
+	</dl>
+</details>
+
+## Objective
+
+_What is this proposal trying to achieve?_
+
+Describe the use cases and requirements for _expressions_
+that enclose parts of a pattern
+and develop a design that satisfies these needs.
+
+## Background
+
+_What context is helpful to understand this proposal?_
+
+## Use-Cases
+
+- Representing markup, such as HTML, as expressions in a pattern
+  rather than as plain character sequences embedded in the _text_
+  of a pattern. This also allows parts of a markup sequence to be
+  programmable.
+
+**_Non-markup use cases to consider (which are not addressed by the design)_**
+
+- Represent templating language constructs so that they can pass through
+  the translation process and are not exposed to translation. For example,
+  Mustache templates.
+
+- Represent translation metadata, such as XLIFF placeholders, so that
+  portions of a message are "protected" or otherwise hinted in the
+  translation process.
+
+## Requirements
+
+Be able to indicate that some identified markup applies to
+a non-empty sequence of pattern parts (text, expressions).
+
+Markup spans may be nested,
+as in `<b>Bold and <i>also italic</i></b>`.
+
+Markup spans may be improperly nested,
+as in `<b>Bold <i>and</b> italic</i>`.
+
+Markup may have localisable options,
+such as `Click <a title="Link tooltip">here</a> to continue`.
+
+As with function options,
+markup options should be defined in a registry so that they can be validated.
+
+## Constraints
+
+Due to segmentation,
+markup may be split across messages so that it opens in one and closes in another.
```

Comment:
I don't think I understand this claim, can you elaborate? Messages are intended to be self-contained entities AFAIK.

Reply:
It is likely that in some cases a user will want to translate something like an entire HTML page via MF2, and rely on automated tooling to separate that into translatable segments, such as sentences, and consider each of them a "message". Such messages could indeed be self-contained from a localization PoV, but not necessarily from a markup PoV.

Reply:
Meh. I don't think this type of segmentation is a good example. A better example might be something that captures some markup that is being used to generate page structure. For example:

```json
"msg1": "{{+section id=$sectionId}{+h2}Hello {$user}{-h2}}",
"msg2": "{{+p}Here is your data:{-p}}",
"msg3": "{+p}Footer{-p}{{-section}}"
```

This would be used to generate the top of a section, with the section close being in a separate resource:

```html
<section id="sectionId"><h2>Hello user</h2>
<p>Here is your data:</p>
<ul>
   <li>... various bits of data generated in code...
</ul>
<p>Footer</p>
</section>
```

The above could be generated in a for-loop over a list of users, for example.

Reply:
Thanks. That is plausible, although I do question whether or not it is _advisable_ because it hinders validation of well-formedness. But at least the motivation for the constraint now makes sense.

Reply:
It is not advisable, since the better course would be to leave the markup in the template. 

Generally, you'd only want inline markup in strings (`span`, `b`, `em`, `a`, etc.). It's harder to think of cases where you'd want to split those up **_and_** retain the markup in the message.